### PR TITLE
feat(connect-form): add kerberos authentication

### PR DESCRIPTION
**Pull request by mcasimir**
Main change:
- add the Kerberos authentication tab

Side changes:
- fix type for query param values (only string should be allowed)
- refactor all connection string manipulation helpers in one place
- move encode/decode password and username to its own helpers as is now reused
- automatically add authSource=$external when selecting authMechanisms that needs it

Screenshot 2022-01-25 at 20 44 14

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "packages-version": "lerna version --allow-branch main --no-push --no-private -m \"chore(release): Bump package versions\"",
     "release": "npm run release --workspace mongodb-compass --",
     "reformat": "lerna run reformat --stream --no-bail",
+    "reformat-changed": "lerna run reformat --stream --no-bail --since origin/HEAD --exclude-dependents",
```
Comment (Anemy) [inline package.json:32]: 🙌

File: packages/connect-form/src/components/advanced-options-tabs/authentication-tab/authentication-default.spec.tsx [hunk 2]
```diff
@@ -156,12 +156,12 @@ describe('AuthenticationDefault Component', function () {
...
     renderComponent({
       connectionStringUrl: new ConnectionStringUrl(
-        'mongodb://C%3BIb86n5b8%7BAnExew%5BTU%25XZy%2C)E6G!dk:password@outerspace:12345'
+        'mongodb://username:C%3BIb86n5b8%7BAnExew%5BTU%25XZy%2C)E6G!dk@outerspace:12345'
       ),
       updateConnectionFormField: updateConnectionFormFieldSpy,
     });
 
-    expect(screen.getByLabelText('Username').getAttribute('value')).to.equal(
+    expect(screen.getByLabelText('Password').getAttribute('value')).to.equal(
```
Comment (Anemy) [inline packages/connect-form/src/components/advanced-options-tabs/authentication-tab/authentication-default.spec.tsx:164]: Ah thanks for catching.

File: packages/connect-form/src/components/advanced-options-tabs/authentication-tab/authentication-gssapi.tsx [hunk 3]
```diff
@@ -1,9 +1,105 @@
...
+          }: React.ChangeEvent<HTMLInputElement>) => {
+            updateConnectionFormField({
+              type: 'update-username',
+              username: value,
+            });
+          }}
+          label="Principal"
+          errorMessage={kerberosPrincipalError}
+          state={kerberosPrincipalError ? 'error' : undefined}
+          value={principal || ''}
```
Comment (Anemy) [inline packages/connect-form/src/components/advanced-options-tabs/authentication-tab/authentication-gssapi.tsx:53]: Previously we had an info link to https://docs.mongodb.com/manual/core/kerberos/#principals
on this input:
Screen Shot 2022-01-25 at 3 57 22 PM
Do we want to keep it?

**Review by Anemy — APPROVED** [review]
lgtm! one question on documentation links

File: packages/connect-form/src/components/advanced-options-tabs/authentication-tab/authentication-gssapi.tsx (hunk 3 above)
Comment (mcasimir) [reply packages/connect-form/src/components/advanced-options-tabs/authentication-tab/authentication-gssapi.tsx:53]: Ah that's a good one, yeah, i'll merge this and add it right away

**Pull request merged**